### PR TITLE
llvmPackages: 17.0.6 -> 18.1.5 on Linux

### PR DESCRIPTION
**Pull request by alyssais**
## Description of changes

The next staging cycle will likely have Rust using LLVM 18 (https://github.com/NixOS/nixpkgs/pull/309580), so to avoid ending up with more LLVMs in closures than necessary, it's time to start working on upgrading the default in general.

I've built all my systems with LLVM 18 with no build regressions (which is better than [16 -> 17](https://github.com/NixOS/nixpkgs/pull/285333), where there was 1), so hopefully it'll just be a few leaf packages that need to have their LLVM versions pinned.  Or maybe we'll be lucky, and the pinning from last round will mean there aren't any?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -16269,7 +16269,7 @@ with pkgs;
       /**/ if platform.isDarwin then 16
       else if platform.isFreeBSD then 16
       else if platform.isAndroid then 12
-      else if platform.isLinux then 17
+      else if platform.isLinux then 18
       else if platform.isWasm then 16
       # For unknown systems, assume the latest version is required.
```
Comment (Smaug123) [inline pkgs/top-level/all-packages.nix:16274]: (This comment is now more obviously out of sync than it was - maybe worth a drive-by relaxing the comment to "assume a recent version", or else raising an issue to track an upgrade? I'm not familiar with how we manage LLVM)

Comment (RossComputerGuy) [reply pkgs/top-level/all-packages.nix:16274]: Could you please elaborate? I'm not sure I understand what your review means.

Comment (Smaug123) [reply pkgs/top-level/all-packages.nix:16274]: The comment says "assume the latest version", and then uses version 17 which is not the latest version. This PR upgrades one of the cases to 18, which (being greater than 17) makes the mismatch between stated intent and actual fact much more obvious.

**Review by RossComputerGuy — COMMENTED** [review]
I see we use `else 17`, is there a reason why we don't change it to 18 or are we only sticking to updating Linux to 18?

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (RossComputerGuy) [reply pkgs/top-level/all-packages.nix:16274]: Yeah well, in the PR title, it states that this is only a Linux upgrade. I'm not sure why we don't specify `"latest"` instead. However the likely reason is because LLVM is a core package and one bad change that is untested could break a lot. I'm sure with how long LLVM 18 has been out, it is okay to change.

Comment (rrbutani) [reply pkgs/top-level/all-packages.nix:16274]: @Smaug123 #312617 bumps the LLVM version for other platforms to `18`.

@RossComputerGuy I believe we can't use `llvmPackages_latest` here because it's an alias:
https://github.com/NixOS/nixpkgs/blob/454d208c022ea7b3fe9007dafaf3a617f74fcd44/pkgs/top-level/aliases.nix#L1418

**Review by RossComputerGuy — APPROVED** [review]

**Comment by RossComputerGuy** [comment]
Tried a `nixpkgs-review` (big mistake with all the `nix log` errors) and a lot of packages are borked on `aarch64-linux`.

```
900 packages marked as broken and skipped
23748 packages failed to build
```

**Comment by RossComputerGuy** [comment]
I tried on the aarch64 builder and it appears that works better. Must be Apple Silicon related weirdness.

**Comment by alyssais** [comment]
Builds fine for me on Apple Silicon, fwiw.

**Comment by alyssais** [comment]
If you only tried building with nixpkgs-review, it was very probably the CPU or memory contention issues I mentioned before.
